docs/xdsl-introduction.ipynb

```diff
@@ -0,0 +1,937 @@
+    "- The `scf` (structured control flow) dialect groups together \n",
+    "    - Operations such as: for/while loops, reduce, yield, if/else statements\n",
+    "    - Attributes such as: no new attributes in this dialect! It instead uses attributes/types from _other_ dialects\n",
+    "- The `arith` (arithmetic) dialect groups together\n",
+    "    - Operation such as: comparison, int/floating point addition/multiplication, or/xor/and logical bitwise operations, and so on\n",
+    "    - Attributes such as: predicates to decide the mode of comparison, flags to determine mode of floating point computation used\n",
+    "- The `async` (asynchornous) dialect groups together operations commonly seen in async applications, as well as associated attributes (types).\n",
+    "    - Operations such as: asynchronous calls, waits, async functions\n",
+    "    - Attributes such as: coroutine reference types, future types\n",
+    "\n",
+    "Dialects can be _combined together_, such that the combined dialect have the operations and attributes from all constituent dialects."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3482c059-bbb4-40f4-8ad5-c8a394431d5d",
+   "metadata": {},
+   "source": [
+    "### Rewrite Patterns\n",
+    "Rewrite pattern is a tool to transform IRs defined using xDSL. \n",
+    "\n",
+    "It makes it easy to specify peephole optimisations. Rewrite patterns specifies a _source_ pattern, and a transformation, which applies itself to the IR by:\n",
+    "- matching a _source_ type pattern within the IR\n",
+    "- updating the IR by applying the predefined transformation\n",
+    "\n",
+    "Multiple rewrite patterns can be applied and composed together in compiler passes.\n",
+    "\n",
+    "A function signature is worth a thousand words... :)\n",
+    "\n",
+    "```python\n",
+    "class RewritePattern(ABC):\n",
+    "    \"\"\"\n",
+    "    A side-effect free rewrite pattern matching on a DAG.\n",
+    "    \"\"\"\n",
+    "\n",
+    "    @abstractmethod\n",
+    "    def match_and_rewrite(self, op: Operation, rewriter: PatternRewriter):\n",
+    "        \"\"\"\n",
+    "        Match an operation, and optionally perform a rewrite using the rewriter.\n",
+    "        \"\"\"\n",
+    "        # here, determine if the operation matches our pattern\n",
+    "        # and define a transformation to be applied on the IR making use of the \n",
+    "        # `rewrite` as an API for changing the IR.\n",
+    "        ...\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7cb89cc7-12de-418b-8a7b-b652ba5a69b8",
+   "metadata": {},
+   "source": [
+    "# xDSL Examples"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b00a9cf1-c2d6-4b7a-a51b-a99f728ab710",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## 1. A SQL dialect (adapted from & also see database_example.ipynb)\n",
```

Comment:
Do we want to include this in the introduction?
We could e.g. put links here to separate notebooks instead. (Those work in JupyterLite too, last time I checked)
We could then have a list of different examples easily skimmed through for users to choose an example similar to their use-case?
What do others think?